Freeze fuse two mms

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #111233
* __->__ #111232

Improves llama_v2 perf locally from 1.48x -> 1.55x. 

A good future rewrite would be to unify the freezing batching with the other batching rules that @yanboliang & co were working on. I want to wait for the forthcoming pre-dispatch changes to settle down first.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

eellison committed:
Freeze fuse two mms

[ghstack-poisoned]

eellison committed:
Update on "Freeze fuse two mms"


Improves llama_v2 perf locally from 1.48x -> 1.55x. 

A good future rewrite would be to unify the freezing batching with the other batching rules that yanboliang & co were working on. I want to wait for the forthcoming pre-dispatch changes to settle down first.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison committed:
Update on "Freeze fuse two mms"


Improves llama_v2 perf locally from 1.48x -> 1.55x. 

A good future rewrite would be to unify the freezing batching with the other batching rules that yanboliang & co were working on. I want to wait for the forthcoming pre-dispatch changes to settle down first.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Chillee commented:
batching rules you say??

eellison commented:
To be more clear, this batches horizontal mms that share a single input where the other inputs are parameters, and constant folds the concat of the params in freezing.

jansel commented on test/inductor/test_inductor_freezing.py:
```diff
@@ -174,6 +174,16 @@ def __init__(self):
+            def __init__(self):
+                super().__init__()
+
+                self.t1 = torch.nn.Parameter(torch.rand(10, 10))
+                self.t2 = torch.nn.Parameter(torch.rand(10, 10))
```
Can you add a test where the sizes of t1/t2 aren't the same?

jansel commented on torch/_inductor/fx_passes/freezing_patterns.py:
```diff
@@ -148,6 +150,24 @@ def matmul_replacement(inp, w1, w2, w3):
+    def matmul_replacement_two(inp, w1, w2):
+        cat_t = torch.cat((w1, w2), dim=1)
+        mm = inp @ cat_t
+        return mm.chunk(2, dim=1)
```
I think this assumes the sizes are equal.

eellison replied:
That condition is checked above:

https://github.com/pytorch/pytorch/blob/bfccf138b02bc4093d5990abdaac71c487842805/torch/_inductor/fx_passes/freezing_patterns.py#L130-L133

jansel replied:
Ah, it should be possible to handle the mismatched case too with a slice.

eellison replied:
will leave that for a follow up

jansel replied:
Sounds good, do add the test though.

eellison committed:
Update on "Freeze fuse two mms"


Improves llama_v2 perf locally from 1.48x -> 1.55x. 

A good future rewrite would be to unify the freezing batching with the other batching rules that yanboliang & co were working on. I want to wait for the forthcoming pre-dispatch changes to settle down first.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison commented:
@pytorchbot merge
